// @line/headless-inspector-cdp/dist/highlight.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Highlight = void 0;
const px = (n = 0) => `${n}px`;
const h = (id, tag, style) => {
    const dom = document.createElement(tag);
    if (id) {
        dom.id = id;
    }
    Object.entries(style).forEach(([key, value]) => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        dom.style[key] = value;
    });
    return dom;
};
const HIGHLIGHT_CONTAINER_ID = 'HIGHLIGHT_CONTAINER_ID';
class Highlight {
    constructor(append) {
        this.node = h(HIGHLIGHT_CONTAINER_ID, 'div', {
            display: 'none',
            position: 'absolute',
        });
        append(this.node);
    }
    hide() {
        this.node.style.display = 'none';
    }
    show(target, { contentColor }) {
        const rect = target.getBoundingClientRect();
        this.node.style.backgroundColor = contentColor;
        this.node.style.width = px(rect.width);
        this.node.style.height = px(rect.height);
        this.node.style.left = px(rect.left);
        this.node.style.top = px(rect.top);
        this.node.style.display = 'block';
    }
}
exports.Highlight = Highlight;
