// @line/headless-inspector/dist/SocketMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocketMap = void 0;
class SocketPool {
    constructor() {
        this.devtool = new Set();
        this.client = new Set();
    }
    add(key, data) {
        this[key].add(data);
    }
    delete(key, data) {
        this[key].delete(data);
    }
}
class SocketMap {
    constructor() {
        this.map = new Map();
    }
    set(key, socket, type) {
        const pool = this.map.get(key);
        if (pool) {
            pool.add(type, socket);
            return;
        }
        const socketPool = new SocketPool();
        socketPool.add(type, socket);
        this.map.set(key, socketPool);
    }
    get(key) {
        return this.map.get(key);
    }
    remove(key, socket, type) {
        const pool = this.map.get(key);
        if (!pool) {
            return;
        }
        pool.delete(type, socket);
    }
    delete(key) {
        this.map.delete(key);
    }
}
exports.SocketMap = SocketMap;
